csharp/ql/lib/semmle/code/csharp/dataflow/internal/ContentDataFlow.qll

```diff
@@ -1,185 +1,215 @@
-     * being the top of the stack), then there is flow from `source.Field2.Field1`.
-     *
-     * For the sink access path, `sinkAp`, the top of the stack represents the content
-     * that was last stored into. That is, if `sinkAp` is `Field1.Field2` (with `Field1`
-     * being the top of the stack), then there is flow into `sink.Field1.Field2`.
-     */
-    final predicate hasFlow(
-      Node source, AccessPath sourceAp, Node sink, AccessPath sinkAp, boolean preservesValue
-    ) {
-      exists(DF::PathNode pathSource, DF::PathNode pathSink |
-        this.(ConfigurationAdapter).hasFlowPath(pathSource, pathSink) and
-        nodeReaches(pathSource, TAccessPathNil(), TAccessPathNil(), pathSink, sourceAp, sinkAp) and
-        source = pathSource.getNode() and
-        sink = pathSink.getNode()
-      |
-        pathSink.getState().(InitState).decode(preservesValue)
-        or
-        pathSink.getState().(ReadState).decode(_, preservesValue)
-        or
-        pathSink.getState().(StoreState).decode(_, preservesValue)
-      )
+    predicate isBarrier(DataFlow::Node node, FlowState state) { none() }
+
+    DataFlow::FlowFeature getAFeature() { result = ContentConfig::getAFeature() }
+
+    // needed to record reads/stores inside summarized callables
+    predicate includeHiddenNodes() { any() }
+  }
+
+  private module Flow = DataFlow::GlobalWithState<FlowConfig>;
+
+  import Flow
+
+  /**
+   * Holds if data stored inside `sourceAp` on `source` flows to `sinkAp` inside `sink`
+   * for this configuration. `preservesValue` indicates whether any of the additional
+   * flow steps defined by `isAdditionalFlowStep` are needed.
+   *
+   * For the source access path, `sourceAp`, the top of the stack represents the content
+   * that was last read from. That is, if `sourceAp` is `Field1.Field2` (with `Field1`
+   * being the top of the stack), then there is flow from `source.Field2.Field1`.
+   *
+   * For the sink access path, `sinkAp`, the top of the stack represents the content
+   * that was last stored into. That is, if `sinkAp` is `Field1.Field2` (with `Field1`
+   * being the top of the stack), then there is flow into `sink.Field1.Field2`.
+   */
+  additional predicate flow(
+    DataFlow::Node source, AccessPath sourceAp, DataFlow::Node sink, AccessPath sinkAp,
+    boolean preservesValue
+  ) {
+    exists(Flow::PathNode pathSource, Flow::PathNode pathSink |
+      Flow::flowPath(pathSource, pathSink) and
+      nodeReaches(pathSource, TAccessPathNil(), TAccessPathNil(), pathSink, sourceAp, sinkAp) and
+      source = pathSource.getNode() and
+      sink = pathSink.getNode()
+    |
+      pathSink.getState().(InitState).decode(preservesValue)
+      or
+      pathSink.getState().(ReadState).decode(_, preservesValue)
+      or
+      pathSink.getState().(StoreState).decode(_, preservesValue)
+    )
+  }
+
+  private newtype TState =
+    TInitState(boolean preservesValue) { preservesValue in [false, true] } or
```

Comment:
Consider using `Boolean` from `codeql.util.Boolean` to avoid having to binding. Same below.